src/util/macros.rs

```diff
@@ -758,30 +772,153 @@ macro_rules! impl_transitive_transmute_from {
+        unsafe impl<T: ?Sized, U: ?Sized> SizeEq<$src<T>> for $dst<U>
+        where
+            T: KnownLayout<PointerMetadata = usize>,
+            U: KnownLayout<PointerMetadata = usize>,
+        {
+            fn cast_from_raw(src: PtrInner<'_, $src<T>>) -> PtrInner<'_, Self> {
+                // SAFETY: `crate::layout::cast_from_raw` promises to satisfy
+                // the safety invariants of `SizeEq::cast_from_raw`, or to
+                // generate a PME. Since `$src<T>` and `$dst<U>` are
+                // `#[repr(transparent)]` wrappers around `T` and `U`
+                // respectively, a `cast_from_raw` impl which satisfies the
+                // conditions for casting from `NonNull<T>` to `NonNull<U>` also
+                // satisfies the conditions for casting from `NonNull<$src<T>>`
+                // to `NonNull<$dst<U>>`.
+
+                // SAFETY: By the preceding safety comment, this cast preserves
+                // referent size.
+                let src: PtrInner<'_, T> = unsafe { cast!(src) };
+                let dst: PtrInner<'_, U> = crate::layout::cast_from_raw(src);
+                // SAFETY: By the preceding safety comment, this cast preserves
+                // referent size.
+                unsafe { cast!(dst) }
+            }
+        }
+
+        if 1 == 0 {
+            let ptr = <$t as KnownLayout>::raw_dangling();
+            #[allow(unused_unsafe)]
+            // SAFETY: This call is never executed.
+            let ptr = unsafe { crate::pointer::PtrInner::new(ptr) };
+            #[allow(unused_unsafe)]
+            // SAFETY: This call is never executed.
+            let ptr = unsafe { cast!(ptr) };
+            let _ = <$dst<$u> as SizeEq<$src<$t>>>::cast_from_raw(ptr);
+        }
+
+
+        impl_for_transmute_from!(T: ?Sized + TryFromBytes => TryFromBytes for $src<T>[<T>]);
+        impl_for_transmute_from!(T: ?Sized + FromBytes => FromBytes for $src<T>[<T>]);
+        impl_for_transmute_from!(T: ?Sized + FromZeros => FromZeros for $src<T>[<T>]);
+        impl_for_transmute_from!(T: ?Sized + IntoBytes => IntoBytes for $src<T>[<T>]);
+
+        impl_for_transmute_from!(U: ?Sized + TryFromBytes => TryFromBytes for $dst<U>[<U>]);
+        impl_for_transmute_from!(U: ?Sized + FromBytes => FromBytes for $dst<U>[<U>]);
+        impl_for_transmute_from!(U: ?Sized + FromZeros => FromZeros for $dst<U>[<U>]);
+        impl_for_transmute_from!(U: ?Sized + IntoBytes => IntoBytes for $dst<U>[<U>]);
+
+        // SAFETY: `$src<T>` is a `#[repr(transparent)]` wrapper around `T`, and
+        // so permits interior mutation exactly when `T` does.
+        unsafe_impl!(T: ?Sized + Immutable => Immutable for $src<T>);
+
+        // SAFETY: `$dst<T>` is a `#[repr(transparent)]` wrapper around `T`, and
+        // so permits interior mutation exactly when `T` does.
+        unsafe_impl!(T: ?Sized + Immutable => Immutable for $dst<T>);
+
+        $blk
```

Comment:
Wonder if the outer `unsafe` required to currently call this macro leaks into cover `$blk`.